File: integration-tests/TestAPI/bindings/c-sharp/Cricut.TestAPI.Tests/ThreadingTests.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using Xunit;

namespace Cricut.TestAPI.Tests {
    [Collection("root")]
    public class ThreadingTests {
        [Fact]
        async Task RunConcurrentJobsInterleavesJobsAndReportsTheFullChunkCount() {
            const int jobCount = 4;
            const int chunksPerJob = 5;
            const int expectedTotal = jobCount * chunksPerJob;

            var observed = new List<(nint jobIndex, nint total)>();
            var gate = new object();
            await Threading.RunConcurrentJobs(jobCount, chunksPerJob, async (jobIndex, total) => {
                await Task.Yield();
                lock (gate) {
                    observed.Add((jobIndex, total));
                }
            });

            Assert.Equal(expectedTotal, observed.Count);

            var perJob = new Dictionary<nint, int>();
            foreach (var (jobIndex, _) in observed) {
                perJob[jobIndex] = perJob.GetValueOrDefault(jobIndex) + 1;
            }
            Assert.Equal(jobCount, perJob.Count);
            foreach (var count in perJob.Values) {
                Assert.Equal(chunksPerJob, count);
            }

            // Concurrent jobs interleave: a serial execution would yield exactly
            // `jobCount - 1` adjacent-index transitions (one per job boundary).
            // Real parallelism produces strictly more.
            var transitions = 0;
            for (var i = 1; i < observed.Count; i++) {
                if (observed[i].jobIndex != observed[i - 1].jobIndex) transitions++;
            }
            Assert.True(transitions > jobCount - 1, $"expected interleaving (>{jobCount - 1} transitions) but saw {transitions}");
        }

        [Fact]
        async Task ComputeRunsInParallelWithTheCallingThread() {
            Assert.True(await Threading.ProveParallelism());
        }

        [Fact]
        async Task MainThreadGenuinelySuspendsOnACondvarWait() {
            Assert.True(await Threading.ProveMainThreadCanPark());
        }
    }
}

File: integration-tests/TestAPI/bindings/kotlin/src/test/kotlin/com/cricut/testapi/ThreadingTests.kt
package com.cricut.testapi

import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.util.Collections
import kotlin.time.Duration.Companion.seconds

internal class ThreadingTests {
    @Test
    fun runConcurrentJobsInterleavesJobsAndReportsTheFullChunkCount() = runTest(timeout = 10.seconds) {
        val jobCount = 4L
        val chunksPerJob = 5L
        val expectedTotal = (jobCount * chunksPerJob).toInt()

        // Callback may be invoked from worker threads; synchronize collection.
        val observed = Collections.synchronizedList(mutableListOf<Pair<Long, Long>>())
        Threading.runConcurrentJobs(jobCount, chunksPerJob) { jobIndex, total ->
            observed.add(jobIndex to total)
        }

        assertEquals(expectedTotal, observed.size)

        val perJob = mutableMapOf<Long, Int>()
        for ((jobIndex, _) in observed) {
            perJob[jobIndex] = (perJob[jobIndex] ?: 0) + 1
        }
        assertEquals(jobCount.toInt(), perJob.size)
        for (count in perJob.values) {
            assertEquals(chunksPerJob.toInt(), count)
        }

        // Concurrent jobs interleave: a serial execution would yield exactly
        // `jobCount - 1` adjacent-index transitions (one per job boundary).
        // Real parallelism produces strictly more.
        var transitions = 0
        for (i in 1 until observed.size) {
            if (observed[i].first != observed[i - 1].first) transitions++
        }
        assertTrue(
            transitions > jobCount - 1,
            "expected interleaving (>${jobCount - 1} transitions) but saw $transitions",
        )
    }

    @Test
    fun computeRunsInParallelWithTheCallingThread() = runTest(timeout = 10.seconds) {
        assertTrue(Threading.proveParallelism())
    }

    @Test
    fun mainThreadGenuinelySuspendsOnACondvarWait() = runTest(timeout = 10.seconds) {
        assertTrue(Threading.proveMainThreadCanPark())
    }
}
